vnext/Chakra/ChakraBytecodeStore.cpp

```diff
@@ -0,0 +1,130 @@
+#pragma pack(pop)
+
+const wchar_t *BytecodeScratchSubdirectory = L"Bytecode";
+const wchar_t *BytecodeFileExtension = L".bytecode";
+
+ChakraBytecodeStore::ChakraBytecodeStore(const std::filesystem::path &executorScratchDirectory) {
+  m_bytecodeDirectory = executorScratchDirectory / BytecodeScratchSubdirectory;
+}
+
+std::unique_ptr<JSBigString> ChakraBytecodeStore::tryObtainCachedBytecode(const std::string &scriptUrl) {
+  auto bytecodePath = bytecodePathFromScriptUrl(scriptUrl);
+  if (!bytecodePath) {
+    return nullptr;
+  }
+
+  auto bytecodeBodyStr = std::make_unique<FileMappingBigString>(bytecodePath->string(), sizeof(BytecodeFileHeader));
+  if (!bytecodeBodyStr->file_data()) {
+    return nullptr;
+  }
+
+  // file_data will go backwards to the beginning of the file instead of going to the specified offset
+  auto header = reinterpret_cast<BytecodeFileHeader *>(bytecodeBodyStr->file_data());
+
+  ChakraVersionInfo curChakraVersion;
+  if (!curChakraVersion.initialize() || header->chakraVersion != curChakraVersion) {
+    return nullptr;
+  }
+
+  folly::Optional<uint64_t> scriptChecksum = calculateScriptChecksum(scriptUrl);
+  if (!scriptChecksum || header->scriptChecksum != scriptChecksum) {
+    return nullptr;
+  }
+
+  return bytecodeBodyStr;
+}
+
+void ChakraBytecodeStore::persistBytecode(const std::string &scriptUrl, const JSBigString &bytecode) {
+  ChakraVersionInfo versionInfo;
+  if (!versionInfo.initialize()) {
+    return;
+  }
+
+  auto bytecodePath = bytecodePathFromScriptUrl(scriptUrl);
+  if (!bytecodePath) {
+    return;
+  }
+
+  folly::Optional<uint64_t> scriptChecksum = calculateScriptChecksum(scriptUrl);
+  if (!scriptChecksum) {
+    return;
+  }
+
+  // bytecode is a binary representation, so we need to pass in the "b" flag to fopen_s
+  FILE *bytecodeFilePtr;
+  if (_wfopen_s(&bytecodeFilePtr, bytecodePath->c_str(), L"wb")) {
+    return;
+  }
+  std::unique_ptr<FILE, decltype(&fclose)> bytecodeFilePtrWrapper(bytecodeFilePtr, fclose);
+
+  // Write a junk header at first in case we fail
+  BytecodeFileHeader junkHeader{};
+
+  if (fwrite(&junkHeader, sizeof(BytecodeFileHeader), 1, bytecodeFilePtrWrapper.get()) != sizeof(BytecodeFileHeader) ||
+      fwrite(bytecode.c_str(), bytecode.size(), 1, bytecodeFilePtrWrapper.get()) != bytecode.size() ||
+      fflush(bytecodeFilePtrWrapper.get())) {
+    return;
+  }
+
+  // Write the real header now
+  BytecodeFileHeader header{CurrentBytecodeFileFormatVersion, *scriptChecksum, versionInfo};
+
+  fseek(bytecodeFilePtrWrapper.get(), 0, SEEK_SET);
+  if (!fwrite(&header, sizeof(BytecodeFileHeader), 1, bytecodeFilePtrWrapper.get()) ||
+      fflush(bytecodeFilePtrWrapper.get())) {
+    return;
+  }
+}
+
+folly::Optional<std::filesystem::path> ChakraBytecodeStore::bytecodePathFromScriptUrl(const std::string &scriptUrl) {
+  std::filesystem::path filename = std::filesystem::path(scriptUrl).filename();
+  if (filename.empty()) {
```

Comment:
>filename [](start = 6, length = 8)

It's not idea that we rely on the script URL to be a filename, but that's what we were doing before. If this is all throwaway it doesn't really matter, but a simple solution to make this more agnostic could be doing something like just base64 encoding the entire scripturl and using that.

Reply:
For that matter, we could also just store bytecode by checksum, which would simplify some things and help if we ever loaded the same script named differently for instance. I would worry a bit about collisions in that case though.

---
In reply to: [341809190](https://github.com/microsoft/react-native-windows/pull/3584#discussion_r341809190) [](ancestors = 341809190)